PolyEngine/Core/Src/Allocator.hpp

```diff
@@ -4,8 +4,87 @@
 
 namespace Poly {
 
-	//FIXME _aligned_malloc is Visual Studio only
+#if defined(_WIN32)
 	inline void* DefaultAlloc(size_t size) { return _aligned_malloc(size, 16); }
 	inline void* DefaultRealloc(void* ptr, size_t size) { return _aligned_realloc(ptr, size, 16); }
 	inline void DefaultFree(void* ptr) { _aligned_free(ptr); }
-}
\ No newline at end of file
+#elif defined(__GNUC__) || defined(__clang__)
+	namespace Impl {
+		constexpr size_t alignment = 16;
+	}
+
+#include <malloc.h>
+
+	inline void* DefaultAlloc(size_t size) {
+		using namespace Impl;
+		void* fresh = aligned_alloc(alignment, size);
+
+#ifdef __has_feature
+#if __has_feature(memory_sanitizer)
+		//we'll be possibly reading this memory later during reallocation, so write over it to avoid triggering unitialized reads detection
+		size_t diff = malloc_usable_size(fresh) - size;
+		memset(static_cast<uint8_t*>(fresh) + size - diff, 0, diff);
+#endif
+#endif
+		return fresh;
+	}
+
+	inline void* DefaultRealloc(void* ptr, size_t size) {
+		using namespace Impl;
+		//POSIX provides no aligned_realloc, so always do a fresh allocation, then copy :(
+		void* fresh = DefaultAlloc(size);
+
+		if (ptr) {
+			memcpy(fresh, ptr, malloc_usable_size(ptr));
+			free(ptr);
+		}
+
+		return fresh;
+	}
+
+	inline void DefaultFree(void* ptr) {
+		free(ptr);
+	}
+
+#if 0
```

Comment:
Technically malloc_usable_size() should only be used for debugging, which is why I included another implementation that uses manual bookkeeping instead (and uses some additional memory for that). I was hesitant which implementation to choose.